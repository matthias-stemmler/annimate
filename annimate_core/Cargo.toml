[package]
name = "annimate_core"
version = "0.1.0"
edition.workspace = true

[build-dependencies]
cargo_metadata = "0.18"

[dependencies]
csv = "1.0"
graphannis = "3.2.1"
graphannis-core = "3.2.1"
itertools = { workspace = true }
regex = { workspace = true }
serde = { workspace = true }
serde_with = "3.7"
thiserror = { workspace = true }
toml = "0.8"
urlencoding = "2.1"

[dev-dependencies]
indoc = "2.0"
insta = { version = "1.34", features = ["serde"] }
