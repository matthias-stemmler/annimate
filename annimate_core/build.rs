//! Build script for `annimate_core`.

use cargo_metadata::MetadataCommand;

fn main() {
    inject_graphannis_version();
}

fn inject_graphannis_version() {
    let metadata = MetadataCommand::new()
        .exec()
        .expect("Failed to run `cargo metadata`");

    let graphannis_version = metadata
        .packages
        .into_iter()
        .find(|p| p.name == "graphannis")
        .expect("`graphannis` package not found in Cargo metadata`")
        .version
        .to_string();

    println!("cargo::rustc-env=GRAPHANNIS_VERSION={graphannis_version}");
}
