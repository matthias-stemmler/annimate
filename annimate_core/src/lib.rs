use std::io::Write;
use std::path::{Path, PathBuf};

use anno::AnnoKeys;
use corpus::CorpusRef;
use format::export;
use graphannis::corpusstorage::CacheStrategy;
use import::{FilesystemEntity, ImportFormat, ImportableCorpus};
use itertools::Itertools;
use metadata::{CorpusSet, MetadataStorage};
use query::Query;
use serde::Serialize;

mod anno;
mod aql;
mod corpus;
mod error;
mod format;
mod import;
mod metadata;
mod node_name;
mod query;
mod util;
mod version;

pub use anno::{ExportableAnnoKey, ExportableAnnoKeys};
pub use aql::{QueryAnalysisResult, QueryNode, QueryNodes};
pub use error::AnnisExportError;
pub use format::{CsvExportColumn, CsvExportConfig, ExportFormat};
pub use graphannis::corpusstorage::CorpusInfo;
pub use graphannis::graph::AnnoKey;
pub use import::ImportedCorpus;
pub use query::{ExportData, ExportDataAnno, ExportDataText, QueryLanguage};
pub use version::{VersionInfo, VERSION_INFO};

pub struct Storage {
    corpus_storage: graphannis::CorpusStorage,
    metadata_storage: MetadataStorage,
}

impl Storage {
    pub fn from_db_dir<P>(db_dir: P) -> Result<Self, AnnisExportError>
    where
        P: AsRef<Path>,
    {
        let corpus_storage = graphannis::CorpusStorage::with_cache_strategy(
            db_dir.as_ref(),
            CacheStrategy::PercentOfFreeMemory(25.0),
            true,
        )?;

        let metadata_storage = MetadataStorage::from_db_dir(
            &db_dir,
            &corpus_storage
                .list()?
                .into_iter()
                .map(|c| c.name)
                .collect_vec(),
        )?;

        Ok(Self {
            corpus_storage,
            metadata_storage,
        })
    }

    pub fn corpora(&self) -> Result<Corpora, AnnisExportError> {
        let sets = self.metadata_storage.corpus_sets();

        let corpora = self
            .corpus_storage
            .list()?
            .into_iter()
            .sorted_by(|a, b| a.name.cmp(&b.name))
            .map(|c| {
                let included_in_sets = sets
                    .iter()
                    .filter(|(_, CorpusSet { corpus_names })| corpus_names.contains(&c.name))
                    .map(|(name, _)| name.clone())
                    .collect();

                Corpus {
                    name: c.name,
                    included_in_sets,
                }
            })
            .collect();

        Ok(Corpora {
            sets: sets.keys().cloned().collect(),
            corpora,
        })
    }

    pub fn delete_corpus(&self, corpus_name: &str) -> Result<(), AnnisExportError> {
        self.corpus_storage.delete(corpus_name)?;
        self.metadata_storage.update_corpus_sets(|corpus_sets| {
            for CorpusSet { corpus_names } in corpus_sets.values_mut() {
                corpus_names.remove(corpus_name);
            }
        })?;
        Ok(())
    }

    pub fn import_corpora<F>(
        &self,
        paths: Vec<PathBuf>,
        on_status: F,
    ) -> Result<Vec<String>, AnnisExportError>
    where
        F: Fn(ImportStatusEvent),
    {
        let mut imported_corpus_names = Vec::new();

        let importable_corpora = import::find_importable_corpora(paths, |message| {
            on_status(ImportStatusEvent::Message {
                index: None,
                message: message.into(),
            });
        })?;

        on_status(ImportStatusEvent::CorporaFound {
            corpora: importable_corpora.iter().map_into().collect(),
        });

        for (index, corpus) in importable_corpora.into_iter().enumerate() {
            on_status(ImportStatusEvent::CorpusImportStarted { index });

            let result = import::import_corpus(&self.corpus_storage, corpus, |message| {
                on_status(ImportStatusEvent::Message {
                    index: Some(index),
                    message: message.into(),
                });
            });

            match result {
                Ok(imported_corpus) => {
                    imported_corpus_names.push(imported_corpus.imported_name.clone());

                    on_status(ImportStatusEvent::CorpusImportFinished {
                        index,
                        result: ImportCorpusResult::Imported {
                            corpus: imported_corpus,
                        },
                    });
                }
                Err(err) => {
                    on_status(ImportStatusEvent::CorpusImportFinished {
                        index,
                        result: ImportCorpusResult::Failed {
                            message: err.to_string(),
                        },
                    });
                }
            }
        }

        Ok(imported_corpus_names)
    }

    pub fn add_corpora_to_set<S>(
        &self,
        corpus_set_name: &str,
        corpus_names: &[S],
    ) -> Result<(), AnnisExportError>
    where
        S: AsRef<str>,
    {
        if corpus_names.is_empty() {
            return Ok(());
        }

        self.metadata_storage.update_corpus_sets(|corpus_sets| {
            corpus_sets
                .entry(corpus_set_name.into())
                .or_default()
                .corpus_names
                .extend(corpus_names.iter().map(|c| c.as_ref().into()));
        })?;

        Ok(())
    }

    pub fn toggle_corpus_in_set(
        &self,
        corpus_set_name: &str,
        corpus_name: &str,
    ) -> Result<(), AnnisExportError> {
        self.metadata_storage.update_corpus_sets(|corpus_sets| {
            if let Some(CorpusSet { corpus_names }) = corpus_sets.get_mut(corpus_set_name) {
                if !corpus_names.remove(corpus_name) {
                    corpus_names.insert(corpus_name.into());
                }
            }
        })?;

        Ok(())
    }

    pub fn validate_query<S>(
        &self,
        corpus_names: &[S],
        aql_query: &str,
        query_language: QueryLanguage,
    ) -> Result<QueryAnalysisResult<()>, AnnisExportError>
    where
        S: AsRef<str>,
    {
        Ok(aql::validate_query(
            self.corpus_ref(corpus_names),
            aql_query,
            query_language,
        )?)
    }

    pub fn query_nodes(
        &self,
        aql_query: &str,
        query_language: QueryLanguage,
    ) -> Result<QueryAnalysisResult<QueryNodes>, AnnisExportError> {
        Ok(aql::query_nodes(
            &self.corpus_storage,
            aql_query,
            query_language,
        )?)
    }

    pub fn exportable_anno_keys<S>(
        &self,
        corpus_names: &[S],
    ) -> Result<ExportableAnnoKeys, AnnisExportError>
    where
        S: AsRef<str>,
    {
        Ok(AnnoKeys::new(self.corpus_ref(corpus_names)).map(AnnoKeys::into_exportable)?)
    }

    pub fn segmentations<S>(&self, corpus_names: &[S]) -> Result<Vec<String>, AnnisExportError>
    where
        S: AsRef<str>,
    {
        Ok(anno::segmentations(self.corpus_ref(corpus_names))?)
    }

    pub fn export_matches<F, S, W>(
        &self,
        corpus_names: &[S],
        aql_query: &str,
        query_language: QueryLanguage,
        format: ExportFormat,
        mut out: W,
        on_status: F,
    ) -> Result<(), AnnisExportError>
    where
        F: Fn(ExportStatusEvent),
        S: AsRef<str>,
        W: Write,
    {
        let anno_keys = AnnoKeys::new(self.corpus_ref(corpus_names))?;
        let query = Query::new(self.corpus_ref(corpus_names), aql_query, query_language)?;
        let matches = query.find(format.get_export_data().cloned())?;

        on_status(ExportStatusEvent::Found {
            count: matches.total_count(),
        });

        export(
            format,
            matches,
            query.nodes(),
            anno_keys.format(),
            &mut out,
            |progress| on_status(ExportStatusEvent::Exported { progress }),
        )?;

        out.flush()?;

        Ok(())
    }

    fn corpus_ref<'a, S>(&'a self, corpus_names: &'a [S]) -> CorpusRef<'a, S> {
        CorpusRef::new(&self.corpus_storage, corpus_names)
    }
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Corpora {
    corpora: Vec<Corpus>,
    sets: Vec<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Corpus {
    name: String,
    included_in_sets: Vec<String>,
}

#[derive(Debug, Serialize)]
#[serde(
    tag = "type",
    rename_all = "snake_case",
    rename_all_fields = "camelCase"
)]
pub enum ExportStatusEvent {
    Found { count: usize },
    Exported { progress: f32 },
}

#[derive(Clone, Debug, Serialize)]
#[serde(
    tag = "type",
    rename_all = "snake_case",
    rename_all_fields = "camelCase"
)]
pub enum ImportStatusEvent {
    CorporaFound {
        corpora: Vec<ImportCorpus>,
    },
    CorpusImportStarted {
        index: usize,
    },
    CorpusImportFinished {
        index: usize,
        result: ImportCorpusResult,
    },
    Message {
        index: Option<usize>,
        message: String,
    },
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportCorpus {
    file_name: String,
    format: ImportFormat,
    trace: Vec<FilesystemEntity<String>>,
}

impl From<&ImportableCorpus> for ImportCorpus {
    fn from(importable_corpus: &ImportableCorpus) -> ImportCorpus {
        ImportCorpus {
            file_name: importable_corpus.file_name.clone(),
            format: importable_corpus.format,
            trace: importable_corpus.trace.clone(),
        }
    }
}

#[derive(Clone, Debug, Serialize)]
#[serde(tag = "type")]
#[serde(rename_all = "snake_case")]
#[serde(rename_all_fields = "camelCase")]
pub enum ImportCorpusResult {
    Imported { corpus: ImportedCorpus },
    Failed { message: String },
}
