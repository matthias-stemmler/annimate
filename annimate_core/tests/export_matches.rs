use std::cell::Cell;
use std::fs;
use std::path::Path;

use annimate_core::{
    AnnimateError, AnnoKey, CsvExportConfig, ExportConfig, ExportData, ExportDataAnno,
    ExportDataText, ExportFormat, ExportStatusEvent, QueryLanguage, Storage, TableExportColumn,
};
use itertools::Itertools;
use serde::Serialize;

const DATA_DIR: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/tests/data");
const DB_DIR: &str = concat!(env!("CARGO_TARGET_TMPDIR"), "/tests/export_matches/db");
const OUTPUT_DIR: &str = concat!(env!("CARGO_TARGET_TMPDIR"), "/tests/export_matches/output");

macro_rules! export_matches_test {
    ($(
        $name:ident: {
            corpus_paths: $corpus_paths:expr,
            corpus_names: $corpus_names:expr,
            aql_query: $aql_query:expr,
            query_language: $query_language:expr,
            export_columns: [$($export_columns:expr,)*],
        }
    )*) => {
        $(
            #[test]
            fn $name() {
                let test_data = TestData {
                    corpus_paths: &$corpus_paths,
                    corpus_names: &$corpus_names,
                    aql_query: $aql_query,
                    query_language: {
                        use QueryLanguage::*;
                        $query_language
                    },
                    export_columns: {
                        use TestTableExportColumn::*;
                        #[allow(unused_imports)]
                        use TestExportData::*;
                        vec![$($export_columns,)*]
                    },
                };

                let db_dir = Path::new(DB_DIR).join(stringify!($name));

                let _ = fs::remove_dir_all(&db_dir);
                let storage = Storage::from_db_dir(db_dir).unwrap();

                fs::create_dir_all(Path::new(OUTPUT_DIR)).unwrap();
                let output_file = Path::new(OUTPUT_DIR).join(concat!(stringify!($name), ".csv"));
                let _ = fs::remove_file(&output_file);

                storage
                    .import_corpora(
                        test_data
                            .corpus_paths
                            .into_iter()
                            .map(|p| Path::new(DATA_DIR).join(p))
                            .collect(),
                        |_| (),
                        || false,
                    )
                    .unwrap();

                storage
                    .export_matches(
                        ExportConfig {
                            corpus_names: test_data.corpus_names.into_iter().cloned().map_into().collect_vec(),
                            aql_query: test_data.aql_query.into(),
                            query_language: test_data.query_language,
                            format: ExportFormat::Csv(CsvExportConfig {
                                columns: test_data
                                    .export_columns
                                    .clone()
                                    .into_iter()
                                    .map_into()
                                    .collect(),
                            }),
                        },
                        &output_file,
                        |_| (),
                        || false,
                    )
                    .unwrap();

                let output = fs::read_to_string(output_file).unwrap();

                insta::with_settings!(
                    {
                         info => &test_data,
                         omit_expression => true,
                    },
                    { insta::assert_snapshot!(output) }
                );
            }
        )*
    };
}

export_matches_test! {
    empty: {
        corpus_paths: ["empty_graphml.zip"],
        corpus_names: ["empty"],
        aql_query: "doc",
        query_language: AQL,
        export_columns: [
            Number,
        ],
    }
    subtok_annos: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "tok=\"tokenized\" | norm=\"subtokenized\"",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("annis", "tok"),
                index: 0,
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "norm"),
                index: 0,
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("structure", "line"),
                index: 0,
            })),
        ],
    }
    subtok_segmentation_tokens: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "pos=\"DT\"",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "lemma"),
                index: 0,
            })),
            Data(Text(TestExportDataText {
                left_context: 4,
                right_context: 4,
                segmentation: None,
                primary_node_indices: None,
            })),
        ],
    }
    subtok_segmentation_diplomatic: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "pos=\"DT\"",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "lemma"),
                index: 0,
            })),
            Data(Text(TestExportDataText {
                left_context: 4,
                right_context: 4,
                segmentation: Some("diplomatic"),
                primary_node_indices: None,
            })),
        ],
    }
    subtok_segmentation_norm: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "pos=\"DT\"",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "lemma"),
                index: 0,
            })),
            Data(Text(TestExportDataText {
                left_context: 4,
                right_context: 4,
                segmentation: Some("norm"),
                primary_node_indices: None,
            })),
        ],
    }
    subtok_gap: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "pos=\"DT\" .5,5 pos=\"DT\"",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "lemma"),
                index: 0,
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "lemma"),
                index: 1,
            })),
            Data(Text(TestExportDataText {
                left_context: 1,
                right_context: 1,
                segmentation: None,
                primary_node_indices: None,
            })),
        ],
    }
    subtok_varying_number_of_match_nodes: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "(pos=\"DT\" .5,5 pos=\"DT\") | pos=\"NN\"",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "pos"),
                index: 0,
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "pos"),
                index: 1,
            })),
            Data(Text(TestExportDataText {
                left_context: 1,
                right_context: 1,
                segmentation: None,
                primary_node_indices: None,
            })),
        ],
    }
    subtok_varying_number_of_match_nodes_with_explicit_primary_node_indices: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "(pos=\"DT\" .5,5 pos=\"DT\") | pos=\"NN\"",
        query_language: AQL,
        export_columns: [
            Data(Text(TestExportDataText {
                left_context: 1,
                right_context: 1,
                segmentation: None,
                primary_node_indices: Some(&[0, 1]),
            })),
        ],
    }
    subtok_optional_nodes: {
        corpus_paths: ["subtok.demo_relANNIS.zip"],
        corpus_names: ["subtok.demo"],
        aql_query: "norm? !.norm,3 norm",
        query_language: AQL,
        export_columns: [
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("grammar", "lemma"),
                index: 0,
            })),
            Data(Text(TestExportDataText {
                left_context: 1,
                right_context: 1,
                segmentation: Some("norm"),
                primary_node_indices: None,
            })),
        ],
    }
    subtok_multiple_corpora: {
        corpus_paths: ["subtok.demo_relANNIS.zip", "subtok.demo2_relANNIS.zip"],
        corpus_names: ["subtok.demo", "subtok.demo2"],
        aql_query: "pos=\"DT\"",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Text(TestExportDataText {
                left_context: 1,
                right_context: 1,
                segmentation: Some("diplomatic"),
                primary_node_indices: None,
            })),
        ],
    }
    subtok_renamed: {
        corpus_paths: ["subtok.demo_renamed_graphml.zip"],
        corpus_names: ["subtok.demo_renamed"],
        aql_query: "pos=\"DT\"",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Text(TestExportDataText {
                left_context: 1,
                right_context: 1,
                segmentation: Some("diplomatic"),
                primary_node_indices: None,
            })),
        ],
    }
    pcc2_primary_node: {
        corpus_paths: ["pcc2_v7_relANNIS.zip"],
        corpus_names: ["pcc2"],
        aql_query: "Sent _i_ NP",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("exmaralda", "Inf-Stat"),
                index: 1,
            })),
            Data(Text(TestExportDataText {
                left_context: 10,
                right_context: 10,
                segmentation: None,
                primary_node_indices: Some(&[1]),
            })),
        ],
    }
    pcc2_primary_nodes_priority: {
        corpus_paths: ["pcc2_v7_relANNIS.zip"],
        corpus_names: ["pcc2"],
        aql_query: "Sent _i_ NP",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Anno(TestExportDataAnno::Corpus {
                anno_key: ("", "language"),
            })),
            Data(Anno(TestExportDataAnno::Document {
                anno_key: ("annis", "doc"),
            })),
            Data(Anno(TestExportDataAnno::MatchNode {
                anno_key: ("exmaralda", "Inf-Stat"),
                index: 1,
            })),
            Data(Text(TestExportDataText {
                left_context: 10,
                right_context: 10,
                segmentation: None,
                primary_node_indices: Some(&[1, 0]),
            })),
        ],
    }
    pcc2_no_context: {
        corpus_paths: ["pcc2_v7_relANNIS.zip"],
        corpus_names: ["pcc2"],
        aql_query: "Sent _i_ NP",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Text(TestExportDataText {
                left_context: 0,
                right_context: 0,
                segmentation: None,
                primary_node_indices: Some(&[]),
            })),
            Data(Text(TestExportDataText {
                left_context: 0,
                right_context: 0,
                segmentation: None,
                primary_node_indices: Some(&[1]),
            })),
            Data(Text(TestExportDataText {
                left_context: 0,
                right_context: 0,
                segmentation: None,
                primary_node_indices: Some(&[0, 1]),
            })),
        ],
    }
    nocoverage: {
        corpus_paths: ["nocoverage.demo_graphml.zip"],
        corpus_names: ["nocoverage.demo"],
        aql_query: "tok=\"no\"",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Text(TestExportDataText {
                left_context: 2,
                right_context: 2,
                segmentation: None,
                primary_node_indices: None,
            })),
        ],
    }
    nondefaultsegmentation: {
        corpus_paths: ["nondefaultsegmentation.demo_graphml.zip"],
        corpus_names: ["nondefaultsegmentation.demo"],
        aql_query: "norm",
        query_language: AQL,
        export_columns: [
            Number,
            Data(Text(TestExportDataText {
                left_context: 2,
                right_context: 2,
                segmentation: Some("norm"),
                primary_node_indices: None,
            })),
        ],
    }
}

#[test]
fn export_cancelled_before_matches_found() {
    let db_dir = Path::new(DB_DIR).join("export_cancelled_before_matches_found");

    let _ = fs::remove_dir_all(&db_dir);
    let storage = Storage::from_db_dir(db_dir).unwrap();

    fs::create_dir_all(Path::new(OUTPUT_DIR)).unwrap();
    let output_file = Path::new(OUTPUT_DIR).join("export_cancelled_before_matches_found.csv");
    let _ = fs::remove_file(&output_file);

    storage
        .import_corpora(
            vec![Path::new(DATA_DIR).join("subtok.demo_relANNIS.zip")],
            |_| (),
            || false,
        )
        .unwrap();

    let result = storage.export_matches(
        ExportConfig {
            corpus_names: vec!["subtok.demo".into()],
            aql_query: "tok".into(),
            query_language: QueryLanguage::AQL,
            format: ExportFormat::Csv(CsvExportConfig {
                columns: vec![TableExportColumn::Number],
            }),
        },
        &output_file,
        |_| (),
        || true,
    );

    assert!(matches!(result, Err(AnnimateError::Cancelled)));
    assert!(!output_file.try_exists().unwrap());
}

#[test]
fn export_cancelled_after_matches_found() {
    let db_dir = Path::new(DB_DIR).join("export_cancelled_after_matches_found");

    let _ = fs::remove_dir_all(&db_dir);
    let storage = Storage::from_db_dir(db_dir).unwrap();

    fs::create_dir_all(Path::new(OUTPUT_DIR)).unwrap();
    let output_file = Path::new(OUTPUT_DIR).join("export_cancelled_after_matches_found.csv");
    let _ = fs::remove_file(&output_file);

    storage
        .import_corpora(
            vec![Path::new(DATA_DIR).join("subtok.demo_relANNIS.zip")],
            |_| (),
            || false,
        )
        .unwrap();

    let cancel_requested = Cell::new(false);

    let result = storage.export_matches(
        ExportConfig {
            corpus_names: vec!["subtok.demo".into()],
            aql_query: "tok".into(),
            query_language: QueryLanguage::AQL,
            format: ExportFormat::Csv(CsvExportConfig {
                columns: vec![TableExportColumn::Number],
            }),
        },
        &output_file,
        |event| {
            if let ExportStatusEvent::Found { .. } = event {
                cancel_requested.set(true);
            }
        },
        || cancel_requested.get(),
    );

    assert!(matches!(result, Err(AnnimateError::Cancelled)));
    assert!(!output_file.try_exists().unwrap());
}

#[derive(Serialize)]
struct TestData {
    corpus_paths: &'static [&'static str],
    corpus_names: &'static [&'static str],
    aql_query: &'static str,
    query_language: QueryLanguage,
    export_columns: Vec<TestTableExportColumn>,
}

#[derive(Clone, Serialize)]
enum TestTableExportColumn {
    Number,
    Data(TestExportData),
}

#[derive(Clone, Serialize)]
enum TestExportData {
    Anno(TestExportDataAnno),
    Text(TestExportDataText),
}

#[derive(Clone, Serialize)]
enum TestExportDataAnno {
    Corpus {
        anno_key: (&'static str, &'static str),
    },
    Document {
        anno_key: (&'static str, &'static str),
    },
    MatchNode {
        anno_key: (&'static str, &'static str),
        index: usize,
    },
}

#[derive(Clone, Serialize)]
struct TestExportDataText {
    left_context: usize,
    right_context: usize,
    segmentation: Option<&'static str>,
    primary_node_indices: Option<&'static [usize]>,
}

impl From<TestTableExportColumn> for TableExportColumn {
    fn from(column: TestTableExportColumn) -> Self {
        match column {
            TestTableExportColumn::Number => TableExportColumn::Number,
            TestTableExportColumn::Data(data) => TableExportColumn::Data(match data {
                TestExportData::Anno(TestExportDataAnno::Corpus {
                    anno_key: (ns, name),
                }) => ExportData::Anno(ExportDataAnno::Corpus {
                    anno_key: AnnoKey {
                        ns: ns.into(),
                        name: name.into(),
                    },
                }),
                TestExportData::Anno(TestExportDataAnno::Document {
                    anno_key: (ns, name),
                }) => ExportData::Anno(ExportDataAnno::Document {
                    anno_key: AnnoKey {
                        ns: ns.into(),
                        name: name.into(),
                    },
                }),
                TestExportData::Anno(TestExportDataAnno::MatchNode {
                    anno_key: (ns, name),
                    index,
                }) => ExportData::Anno(ExportDataAnno::MatchNode {
                    anno_key: AnnoKey {
                        ns: ns.into(),
                        name: name.into(),
                    },
                    index,
                }),
                TestExportData::Text(TestExportDataText {
                    left_context,
                    right_context,
                    segmentation,
                    primary_node_indices,
                }) => ExportData::Text(ExportDataText {
                    left_context,
                    right_context,
                    segmentation: segmentation.map(|s| s.to_string()),
                    primary_node_indices: primary_node_indices.map(Into::into),
                }),
            }),
        }
    }
}
