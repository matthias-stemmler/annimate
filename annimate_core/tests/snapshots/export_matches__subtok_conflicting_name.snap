---
source: annimate_core/tests/export_matches.rs
info:
  corpus_paths:
    - subtok.demo_relANNIS.zip
    - subtok.demo_relANNIS.zip
  corpus_names:
    - subtok.demo
    - subtok.demo (1)
  aql_query: tok
  query_language: AQL
  export_columns:
    - Number
    - Data:
        Anno:
          Corpus:
            anno_key:
              - annis
              - node_name
    - Data:
        Anno:
          Document:
            anno_key:
              - annis
              - node_name
    - Data:
        Anno:
          MatchNode:
            anno_key:
              - annis
              - node_name
            index: 0
---
Number,Corpus node_name,Document node_name,#1 node_name
1,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_1
2,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_2
3,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_3
4,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_4
5,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_5
6,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_6
7,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_7
8,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_8
9,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_9
10,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_10
11,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_11
12,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_1
13,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_2
14,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_3
15,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_4
16,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_5
17,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_6
18,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_7
19,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_8
20,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_9
21,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_10
22,subtok.demo,subtok.demo/subtok.demo.doc01,subtok.demo/subtok.demo.doc01#tok_11
